spark3/src/main/java/org/apache/iceberg/spark/source/SparkMergeScan.java

```diff
@@ -111,7 +111,7 @@ public Statistics estimateStatistics() {
   }
 
   @Override
-  public void filterFiles(Set<String> locations) {
+  public synchronized void filterFiles(Set<String> locations) {
```

Comment:
Since we reuse the same scan instance in two scan relations, it seems safer to add synchronization.

Reply:
Could we also re-use the DynamicFilter plan? Then we would only run `prepare` once and would not need to synchronize

Reply:
Can you elaborate a bit more? There is only one DynamicFilter node right now (the second branch uses a normal scan node). Technically, we don't need to add synchronization to `filterFiles` but the other two cases seem valid.

Reply:
The dynamic filter keeps a reference to the Filterable, which is this. As long as there is only one DynamicFilter, per Scan, then I don't think we really need this. It's fine to add it, but doesn't seem necessary.

I wasn't sure earlier that we reused the dynamic filter. If so, we should be good to go. However, I do think that synchronization on `files()` and `tasks()` is a good idea because those will refresh the instance variable. So maybe adding it here is a good idea, too.

Reply:
I think I am going to revert this part to avoid confusion. I think we are safe without it.

Reply:
Removed. Resolving this.